tests/test_graph/test_parameters.py

```diff
@@ -0,0 +1,87 @@
+"""Tests for graph module."""
+
+import re
+from typing import Literal, TypeAlias
+
+import numpy as np
+import pytest
+
+from causalprog.graph import DistributionNode, Graph, ParameterNode
+
+NormalGraphNodeNames: TypeAlias = Literal["mean", "cov", "outcome"]
+NormalGraphNodes: TypeAlias = dict[
+    NormalGraphNodeNames, DistributionNode | ParameterNode
+]
+
+
+@pytest.mark.parametrize(
+    ("param_values_before", "params_to_set", "expected"),
+    [
+        pytest.param(
+            {},
+            {"outcome": 4.0},
+            TypeError("Node outcome is not a parameter node."),
+            id="Give non-parameter node",
+        ),
+        pytest.param(
+            {},
+            {"mean": 4.0},
+            {"mean": 4.0, "cov": None},
+            id="Set only one parameter",
+        ),
+        pytest.param(
+            {},
+            {},
+            {"mean": None, "cov": None},
+            id="Doing nothing is fine",
+        ),
+        pytest.param(
+            {"mean": 0.0, "cov": 0.0},
+            {"cov": 1.0},
+            {"mean": 0.0, "cov": 1.0},
+            id="Omission preserves current value",
+        ),
+    ],
+)
+def test_set_parameters(
+    normal_graph: Graph,
+    param_values_before: dict[NormalGraphNodeNames, float],
+    params_to_set: dict[str, float],
+    expected: Exception | dict[NormalGraphNodeNames, float],
+) -> None:
+    """Test that we can identify parameter nodes, and set their values."""
+    graph = normal_graph()
+    parameter_nodes = graph.parameter_nodes
+    assert graph.get_node("mean") in parameter_nodes
+    assert graph.get_node("cov") in parameter_nodes
+    assert graph.get_node("outcome") not in parameter_nodes
+
+    # Set any pre-existing values we might want the parameter nodes to have in
+    # this test.
+    for node_label, value in param_values_before.items():
+        n = graph.get_node(node_label)
+        assert isinstance(n, ParameterNode), (
+            "Cannot set .value on non-parameter node (test input error)."
+        )
+        n.value = value
+
+    # Check behaviour of set_parameters method.
+    if isinstance(expected, Exception):
+        with pytest.raises(type(expected), match=re.escape(str(expected))):
```

Comment:
There's now a [`raises_context` fixture](https://github.com/UCL/causalprog/blob/aae377d7517432623750a3c5c3015ba5c76c6735/tests/fixtures/general.py#L20) to save on this bit of code cropping up again and again.

```suggestion
        with raises_context(expected):
```

You'll need to add `raises_context` to the arguments of the test too. And then to stop ruff complaining, you don't need to `import re` anymore either.